Add py-pyproj 2.2.0

adamjstewart opened the pull request:
@citibeth Can you take a look at this? I tried building without the `proj` dependency but it crashed saying the system installation was too old. When I added a dependency on `proj`, everything built fine.

citibeth commented on var/spack/repos/builtin/packages/py-pyproj/package.py:
```diff
@@ -9,19 +9,18 @@
-    # The changes in this "version" fix some bugs, especially with Python3 use.
-    version('1.9.5.1.1', commit='0be612cc9f972e38b50a90c946a9b353e2ab140f')
-    version('1.9.5.1', 'a4b80d7170fc82aee363d7f980279835')
+    maintainers = ['adamjstewart']
```
Given the issues that came up here, please list `citibeth` as the maintainer.  In general... when a `maintainer` is added to a package that doesn't previously have such a designation... please consider using `git blame` to see who originally built the package.

adamjstewart replied:
I don't like adding people as maintainers without their consent, but I'll add you. Let me know if you want to be added to any other packages.

citibeth replied:
I think it's reasonable that if the large majority of work on a package was done by one person (i.e. it was created by that person and never substantially updated by anyone else); then the original package author should be given the chance to be a maintainer.

adamjstewart commented:
> I rely on the unofficial 1.9.5.1.1 release, which includes fixes for Python3. This PR replaces that with 2.2.0, which probably includes the fixes from 1.9.5.1.1. However, it also requires PROJ.6, which my software is not prepared to deal with.

It looks like there is an official 1.9.6 release. I can also add any of the 2.0:2.1 releases that still support PROJ.4 if you want: https://pypi.org/project/pyproj/#history. What do you think? I think an official release would be better than a particular commit.

adamjstewart commented:
> Also, why is py-aenum required? I've been running without it on my (Spack-installed) python@3.5 for a while now.

Good catch. It looks like the `aenum` dependency was added in version 2.2.0: https://github.com/pyproj4/pyproj/pull/302

I'll update that.

citibeth commented:
> It looks like there is an official 1.9.6 release. I can also add any of the 2.0:2.1 releases that still support PROJ.4 if you want

Do you know the difference between v1 and v2?  In any case, I think the latest releases of `1.9`, `2.0` and `2.1` would be worthwhile.

>  I think an official release would be better than a particular commit.

Agreed.  If `1.9.6` is added, we can remove the unofficial release.  (I'm not going to bother testing for now, I'll just assume it works).




Agreed

citibeth commented:
Have you gotten to the bottom of whether `proj` is a depenency, or whether it's included (and compiled) as part of the `py-pyproj` build; and whether this has changed over time with different versions of `py-pyproj`?  Obviously the former is better, if `py-pyproj` supports it.

adamjstewart commented:
Okay, I added @citibeth as a maintainer, added the latest official release that includes Python 3 support but doesn't require PROJ.6, restricted the `py-aenum` dependency, and made sure that the build always uses the Spack `proj` installation instead of a system installation or bundled copy. That way we'll always end up with the same version of `proj` in the DAG. Let me know if there is anything else you want me to fix.

adamjstewart commented:
> Have you gotten to the bottom of whether proj is a depenency

It is a dependency

> or whether it's included (and compiled) as part of the py-pyproj build

It is included, but it crashed when I tried building on a system that already had an older version of `proj` installed. Perhaps it works on a system with no `proj`?

> and whether this has changed over time with different versions of py-pyproj?

As far as I can tell, this logic has never changed.

citibeth commented on var/spack/repos/builtin/packages/py-pyproj/package.py:
```diff
@@ -9,19 +9,24 @@
-    # The changes in this "version" fix some bugs, especially with Python3 use.
-    version('1.9.5.1.1', commit='0be612cc9f972e38b50a90c946a9b353e2ab140f')
-    version('1.9.5.1', 'a4b80d7170fc82aee363d7f980279835')
+    maintainers = ['adamjstewart', 'citibeth']
```
Can you list me first, since I was the original author?  I don't have a burning need to control all possible changes.  If somebody suggests changes to this package, I'd want to have the chance to review it, in case it inadvertently breaks something I rely upon.

citibeth commented on var/spack/repos/builtin/packages/py-pyproj/package.py:
```diff
@@ -9,19 +9,24 @@
+
+    version('2.2.0',   sha256='0a4f793cc93539c2292638c498e24422a2ec4b25cb47545addea07724b2a56e5')
+    version('1.9.6',   sha256='e0c02b1554b20c710d16d673817b2a89ff94738b0b537aead8ecb2edc4c4487b')
+    version('1.9.5.1', sha256='53fa54c8fa8a1dfcd6af4bf09ce1aae5d4d949da63b90570ac5ec849efaf3ea8')
```
Can we add a restriction that `1.9.5.1` does NOT work with Python3?

citibeth commented on var/spack/repos/builtin/packages/py-pyproj/package.py:
```diff
@@ -9,19 +9,24 @@
     depends_on('py-setuptools', type='build')
+    depends_on('py-cython', type='build')
+    depends_on('py-aenum', type=('build', 'run'), when='@2.2:^python@:3.5')
+    depends_on('proj')
```
Is this required?  (Research needed)

adamjstewart replied:
It looks like the same PR that added PROJ.6 support also removed the vendored copy of PROJ.4?
https://github.com/pyproj4/pyproj/pull/155

citibeth commented on var/spack/repos/builtin/packages/py-pyproj/package.py:
```diff
@@ -9,19 +9,24 @@
-    # which is built internally as part of the py-proj build.
-    # Adding depends_on('proj') will cause mysterious build errors.
+    def setup_environment(self, spack_env, run_env):
+        spack_env.set('PROJ_DIR', self.spec['proj'].prefix)
```
Are we sure this isn't needed for `run_env`?

citibeth replied:
OK, I found the PROJ docs explaining the `PROJ_DIR` variable.  Do you think we should follow all of its instructions on how to run `setup` with an externally compiled `proj`?  That way we best avoid problems with system-supplied PROJ...

https://pyproj4.github.io/pyproj/v2.2.0rel/installation.html#setup-proj

> If you have a previous version of PROJ installed alongside the current version of PROJ (6.1.0), the best way to avoid conflicts is to:
> 1. Remove the previous PROJ from PATH & unset old PROJ_LIB environment variable (temporarily)
> 2. Install PROJ to the internal PROJ directory (pyproj/proj_dir)
> 3. Set the environment variable PROJ_DIR to point to the internal PROJ directory
> 4. Set the environment variable PROJ_WHEEL=true
> 5. Build pyproj

adamjstewart replied:
It would be troublesome to have to install `proj` to `pyproj/proj_dir`

citibeth replied:
> It would be troublesome to have to install `proj` to `pyproj/proj_dir`

I suppose so.  But we can still:
1. Prepend `$PATH` with `spec['proj'].prefix.bin`
2. Set `$PROJ_DIR` as this PR dcurrently does.
3. Set `$PROJ_WHEEL=true`.  I have no idea why, but they seem to like this magic...

adamjstewart replied:
1. Spack automatically adds `spec['proj'].prefix.bin` to the `PATH`.
2. Got this covered
3. We don't want to install wheels with Spack. Maybe these instructions are for use by `pyproj` developers?

adamjstewart commented on var/spack/repos/builtin/packages/py-pyproj/package.py:
```diff
@@ -9,19 +9,26 @@
+    version('1.9.5.1', sha256='53fa54c8fa8a1dfcd6af4bf09ce1aae5d4d949da63b90570ac5ec849efaf3ea8')
+
+    depends_on('python@:2', when='@:1.9.5.1')
+    depends_on('python@3:', when='@2.3:')
```
https://github.com/pyproj4/pyproj/issues/331

citibeth commented:
I'm happy to merge now...

adamjstewart commented:
@citibeth can you also merge the PR that adds PROJ.6 and the PRs that fix the packages that depend on it? As is, this package probably won't work.
